akka-stream/src/main/scala/akka/stream/impl/SetupStage.scala

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.annotation.InternalApi
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+import akka.stream._
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+
+import scala.concurrent.{ Future, Promise }
+
+/** Internal Api */
+@InternalApi private[stream] final class SetupSinkStage[T, M](factory: ActorMaterializer ⇒ Attributes ⇒ Sink[T, M])
+  extends GraphStageWithMaterializedValue[SinkShape[T], Future[M]] {
+
+  private val in = Inlet[T]("SetupSinkStage.in")
+  override val shape = SinkShape(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[M]) = {
+    val matPromise = Promise[M]
+    (createStageLogic(matPromise), matPromise.future)
+  }
+
+  private def createStageLogic(matPromise: Promise[M]) = new GraphStageLogic(shape) {
+    import SetupStage._
+
+    val subOutlet = new SubSourceOutlet[T]("SetupSinkStage")
+    subOutlet.setHandler(delegateToInlet(() ⇒ pull(in), () ⇒ cancel(in)))
+    setHandler(in, delegateToSubOutlet(() ⇒ grab(in), subOutlet))
+
+    override def preStart(): Unit = {
+      val sink = factory(ActorMaterializerHelper.downcast(materializer))(attributes)
+
+      val mat = Source.fromGraph(subOutlet.source).runWith(sink.withAttributes(attributes))(subFusingMaterializer)
+      matPromise.success(mat)
+    }
+  }
+
+}
+
+/** Internal Api */
+@InternalApi private[stream] final class SetupFlowStage[T, U, M](factory: ActorMaterializer ⇒ Attributes ⇒ Flow[T, U, M])
```

Comment:
Any thoughts regarding curried vs non curried function here? We don't often do curried style things in Akka?

Reply:
This is the internal implementation. We could uncurry here and have `factory` be a two argument function. I am fine with either.

What do you think about having the curried `factory` in the user facing API? I find it quite helpful to be able to add `implicit` keywords to the function parameters when constructing the `factory` function. For example:

```scala
Source.setup { implicit mat => _ =>
  Source.single(someStream.run())
}
```

Reply:
ah, good point about being able implicit params. I know also the play action/controller API does (or used to) do it like that.

👍